runners/direct-java/src/test/java/org/apache/beam/runners/direct/DirectRunnerTest.java

```diff
@@ -299,6 +304,35 @@ public void hang(ProcessContext context) throws InterruptedException {
     assertThat(result.getState(), is(State.RUNNING));
   }
 
+  private static final AtomicBoolean TEARS_DOWN_FNS_BEFORE_FINISHING_STATE = new AtomicBoolean();
+
+  @Test
+  public void tearsDownFnsBeforeFinishing() {
+    final Pipeline pipeline = getPipeline();
+    pipeline.apply(Create.of("a"))
+      .apply(ParDo.of(new DoFn<String, String>() {
+        private String name = testName.getMethodName();
+
+        @ProcessElement
+        public void onElement(final ProcessContext ctx) {
+          // no-op
+        }
+
+        @Teardown
+        public void teardown() {
+          try {
+            Thread.sleep(5000);
```

Comment:
You can reduce this sleep pretty significantly, and it should force the race in almost every case - 100 ms is almost definitely sufficient, and 250 should be more than enough

Reply:
Actually not cause:
1. it would depend too much on the machine
2. it will likely never work on windows

This is a "safe" value to error too much false positive failures.

Reply:
Just dropping in here - is there no way to use synchronization / latches to actually force the race 100% of the time with no waiting?

Reply:
yep, all ways involving barriers or equivalents will lead to no more test the user behavior we want to cover here. I went from here and got several false positive. Long story short we wait teardown is done so we must have a teardown which overpasses the "main" thread execution enough to require the main to wait for it. If we sync it we sync the user code but not the outside code which is what we want to test too through this one.

Reply:
Ah, I see what you are trying to test. What about:

DoFn teardown:
 - unlatch "start signal"
 - block on "waiting"
 - set the atomic boolean after

Main:
 - run pipeline
 - block on "start signal"
 - launch async waitToFinish
 - yield
 - unlock "waiting"
 - block on async waitToFinish

There is a tiny chance that "waiting" is unlocked before waitToFinish actually gets called but it doesn't require a 5 second delay.

Another possibility - doesn't have to be this PR but maybe a followup Jira - is to add package-private hooks into DirectPipelineResult so that you can block on something that guarantees waitToFinish has been called.

In any case we need a Jira to track a test with a 5 second sleep.

Reply:
I really like the callback idea. This is what we did in BatchEE and it is quite convenient for tests and even in user code sometimes for very particular and advanced usages a bit outside the default main API. Do you want to log this idea yourself?

Reply:
The asynchronous approach seems to be totally reasonable (in that I don't see any reason it wouldn't work) and much preferable to an extended sleep.

Of note, there's no way to in every case force the `waitUntilFinish` call to happen before the call to `teardown` is made, because that's based on the mercy of the scheduler and the rate at which the pipeline completes.

Reply:
The callback requires its own PR IMHO so for this one I will rewrite a bit the test and just use a sleep(1s) to limit the potential side effects. Should be ok for now.